### ISSUE TITLE
Trouble with basic example

### ISSUE THREAD

**Issue by markau** [sslotsky/violet-paginator#3]
Hi, I am following the example from the README, but getting this error:

`PaginationWrapper.js?8937:37 Uncaught TypeError: Cannot read property 'find' of undefined(…)`

It seems that `state.pagination()` is `undefined`.

Do I need to include a reducer for this library in my store by any chance?

**Comment by sslotsky** [comment]
Yes! This is admittedly missing from the README (my apologies), but it is documented here

https://sslotsky.gitbooks.io/violet-paginator/content/introduction.html

**Comment by sslotsky** [comment]
Just added it to the README, thanks for the nudge!

https://github.com/sslotsky/violet-paginator/commit/06eb8747093f84aaa175398070a4cd8a51d5b73e

I'll keep this issue open until you get your basic example working. That will help illuminate anything that might be lacking in the initial setup guides and allow us to track them in one place.

**Comment by markau** [comment]
hey @sslotsky, thanks.  Adding the reducer means I have pagination buttons appearing on the page, and it attempts to call the fetch action. So that's great progress!

I have an `action creator` being called for the fetch, which calls the API and then passes the response to an `action` - however, I don't know which state key that needs to be inserted into though?  This is probably more a general redux question than anything. My understanding of redux is tenuous at best!

**Comment by sslotsky** [comment]
@markau You actually don't need to dispatch your own action when you receive the response. `violet-paginator` does that for you. You just need to make sure that `violet-paginator` knows how to _read_ the response.

There are multiple ways that you can tell `violet-paginator` how to read the response. The sections on [single list configuration](https://sslotsky.gitbooks.io/violet-paginator/content/single_list_configuration.html) and [global list configuration](https://sslotsky.gitbooks.io/violet-paginator/content/global_configuration.html) should help guide the way.

**Comment by sslotsky** [comment]
Note that if `resp.data` already looks like this when it comes back from the server:

```
results: [{ ... }, { ... }],
total_count: 2
```

then `violet-paginator` will already be able to read it with no configuration necessary. However, if it looks different than that, you'll need to use one of the techniques described in the configuration docs.

**Comment by markau** [comment]
I meant to say that I did RTFM and work some of that out. Thanks!

Postman shows the API response as:

````
{
  "data": {
    "results": [
      {
        "firstName": "bob",
        "lastName": "smith"
      },
      {
        "firstName": "george",
        "lastName": "costanza"
      }
    ],
    "total_count": 2
  }
}
````

In such a case, I believe I should use `resultsProp="results"` (although I suspect results is the default prop name).

I have headers set as:

````
  headers() {
    return [{
      field: 'firstName',
      text: 'First Name'
    }, {
      field: 'lastName',
      text: 'Last Name'
    }]
  }
````

So I think I should be seeing some data. Must be close now! Can you just let me know if you spot an error above, otherwise I will keep debugging!

**Comment by markau** [comment]
btw I am consuming the API with:

````
return fetch(url)
.then(response => response.json())
````

**Comment by sslotsky** [comment]
Your example looks correct to me, and yes, `resultsProp` is set to `"results"` by default so you shouldn't need to override.

As far as your API consumption I'm not familiar with calling `response.json()`. Perhaps this is just something different about the libraries we're using to fetch. Normally I wouldn't attach a `.then` handler at all if the response is already formatted correctly:

```javascript
export function fetch(pageInfo) {
  return fetch(url, pageInfo.query);
}
```

Are you seeing any errors of any sort? Or is your data table simply not displaying your results? Feel free to point me to a repo if you decide you're stuck.

**Comment by markau** [comment]
Thanks mate. I'm sure it's me and not the library. I am using the [wg-fetch](https://github.com/github/fetch) library. The promise is returning the JSON I outlined above, but there must be a problem (it is not displaying results on the page).  Using Redux Devtools I see that this `action` is being dispatched:

`@@violet-paginator/RESULTS_UPDATED_ERROR`

And that:

`state.pagination.0.results = []`

I'll have a break and another look at it. Thanks for your help so far!

**Comment by sslotsky** [comment]
Looking at it again, I'm suspecting that you _might_ need to do this:

```javascript
export function fetchRecords(pageInfo) {
  return () => fetch(url, pageInfo.query).then(resp => resp.json().data)
}
```

**Comment by markau** [comment]
The way the wg-fetch library is loading the data must be the problem - the promise seems to not be resolving. Is there a better library to use to load data? (I can't use native JavaScript `fetch()` because it doesn't seem to work in IE.) I've been perfectly happy with wg-fetch on this project - using it with redux-form for posting data.

I added some debugging to `fetchingComposables.js`:

````
    return promise.then(function (resp) {
      console.log(resp);
      <snip>
    }).catch(function (error) {
      console.log(error);
      <snip>
    });
````

Those two `console.log()` show this respectively:

`undefined`
`TypeError: Cannot read property 'data' of undefined(…)`

To ensure completeness for what I guess is now a bug report :smile: :

The data is being returned from the API via:

````
        return fetch(url)
        .then(function(response) {
          return response.json()
        })
        .then(function(json) {
          console.log('parsed json', json)
        }).catch(function(ex) {
          console.log('parsing failed', ex)
        })
````

Postman shows the response as:

````
{
  "data": {
    "results": [
      {
        "firstName": "bob",
        "lastName": "smith"
      },
      {
        "firstName": "george",
        "lastName": "costanza"
      }
    ],
    "total_count": 2
  }
}
````

The component is implemented on the page as:

````
  headers() {
    return [{
      field: 'firstName',
      text: 'First Name'
    }, {
      field: 'lastName',
      text: 'Last Name'
    }]
  }

  render() {
    const { fetch } = this.props
    console.log(this.props);
    const paginator = (
      <VioletPaginator listId="recipes" fetch={fetch} />
    )

    return (
      <section>
        {paginator}
        <VioletDataTable listId="recipes" fetch={fetch} headers={this.headers()} />
        {paginator}
      </section>
    )
  }
````

**Comment by sslotsky** [comment]
Probably my best bet is to try stepping through your code myself. Do you have a link to a repo that I can clone?

**Comment by markau** [comment]
Ahh yes, of course. I'll try to replicate in a minimal and clean solution that I am free to post publicly (it's fairly deep in a work project.)

**Comment by sslotsky** [comment]
Thanks. Although I'll agree that the fact that your `response` object seems to be `undefined` is highly suspect. Definitely seems like whatever problem you're having is occurring before the data gets to `violet-paginator`

**Comment by markau** [comment]
I tried [a different http library](https://github.com/mzabriskie/axios) and got it working 😀 

It works with both libraries; the necessary step was as you suspected, the api response has to be manipulated so that the promise returns a data structure of `Object {data: Object}`, i.e. `return response.data` not `return response`.  That was simpler to do with Axios so it got me to the right answer faster.  I could also change the API to return an appropriate data structure. I tried all of these things but I guess not in the right combination.

Thanks for your time, and this sweet library!  I can't wait to start using it.
